src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -2236,6 +2231,14 @@ private bool IsTypeDeclarationStart()
                 case SyntaxKind.ExplicitKeyword or SyntaxKind.ImplicitKeyword
                     when this.PeekToken(1).ContextualKind is SyntaxKind.ExtensionKeyword:
 
+                    // If we have `implicit extension E` then this is definitely an extension as it could not be parsed
+                    // as anything else.
+                    if (IsTrueIdentifier(this.PeekToken(2)))
+                        return true;
+
+                    // we have `implicit extension ...` technically this could be something like `implicit
+                    // extension.operator X(` (an implicit conversion with a explicit interface impl name). For compat,
+                    // determine the parsing strategy based on the lang version.
                     return IsFeatureEnabled(MessageID.IDS_FeatureExtensions);
```

Comment:
the policy is this:

in C# 13, we always treat `implicit extension` as the start of an extension type.  Prior to C# 13, we take the approach of ensuring any prior legal parses stay legal.  But if it would not be legal, we treat it as an extension (with an error that they're not available in your version).